src/passes/GlobalSubtyping.cpp

```diff
@@ -0,0 +1,266 @@
+// writes to that field in the entire program allow doing so.
+//
+
+#include "ir/lubs.h"
+#include "ir/struct-utils.h"
+#include "ir/type-updating.h"
+#include "ir/utils.h"
+#include "pass.h"
+#include "support/unique_deferring_queue.h"
+#include "wasm-type.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+// We use a LUBFinder to track field info. A LUBFinder keeps track of the best
+// possible LUB so far. The only extra functionality we need here is whether
+// there is a default null value (which would force us to keep the
+// type nullable).
+struct FieldInfo : public LUBFinder {
+  bool nullDefault = false;
+
+  void noteNullDefault() { nullDefault = true; }
+
+  bool hasNullDefault() { return nullDefault; }
+
+  bool noted() { return LUBFinder::noted() || nullDefault; }
+
+  Type get() {
+    auto ret = LUBFinder::get();
+    if (nullDefault && !ret.isNullable()) {
+      ret = Type(ret.getHeapType(), Nullable);
+    }
+    return ret;
+  }
+
+  bool combine(const FieldInfo& other) {
+    auto old = nullDefault;
+    if (other.nullDefault) {
+      noteNullDefault();
+    }
+    return LUBFinder::combine(other) || nullDefault != old;
+  }
+
+  // Force the lub to a particular type.
+  void set(Type type) { lub = type; }
+
+  void dump(std::ostream& o) {
+    std::cout << "FieldInfo(" << lub << ", " << nullDefault << ")";
+  }
+};
+
+struct FieldInfoScanner : public Scanner<FieldInfo, FieldInfoScanner> {
+  Pass* create() override {
+    return new FieldInfoScanner(functionNewInfos, functionSetGetInfos);
+  }
+
+  FieldInfoScanner(FunctionStructValuesMap<FieldInfo>& functionNewInfos,
+                   FunctionStructValuesMap<FieldInfo>& functionSetGetInfos)
+    : Scanner<FieldInfo, FieldInfoScanner>(functionNewInfos,
+                                           functionSetGetInfos) {}
+
+  void noteExpression(Expression* expr,
+                      HeapType type,
+                      Index index,
+                      FieldInfo& info) {
+    info.note(expr);
+  }
+
+  void
+  noteDefault(Type fieldType, HeapType type, Index index, FieldInfo& info) {
+    // Default values do not affect what the heap type of a field can be turned
+    // into. Note them, however, as they force us to keep the type nullable.
+    if (fieldType.isRef()) {
+      info.noteNullDefault();
+    }
+  }
+
+  void noteCopy(HeapType type, Index index, FieldInfo& info) {
+    // Copies do not add any type requirements at all: the type will always be
+    // read and written to a place with the same type.
+  }
+
+  void noteRead(HeapType type, Index index, FieldInfo& info) {
+    // Nothing to do for a read, we just care about written values.
+  }
+};
+
+struct GlobalSubtyping : public Pass {
+  StructValuesMap<FieldInfo> finalInfos;
+
+  void run(PassRunner* runner, Module* module) override {
+    if (getTypeSystem() != TypeSystem::Nominal) {
+      Fatal() << "GlobalSubtyping requires nominal typing";
+    }
```

Comment:
Basic question: Why does this pass (and `GlobalTypeOptimization` too) require nominal typing?

Reply:
Nominal typing lets us see exactly which types are subtypes of others. Technically we could compute that for structural subtyping too, but it takes a lot more work (we'd need to duplicate the logic in the canonicalization code, I assume), and also it would be less effective anyhow as when it merges types it limits optimization opportunities (since we can only optimize the merged type in one way, and not each of the original types differently as is best for it).